### PR TITLE
feat: add external sort algorithm

### PR DESCRIPTION
**Pull request by bls-lehoai**
#### Description of Change
Implement new algorithm: external sort

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: sorting/external_sort.cpp [hunk 1]
```diff
@@ -0,0 +1,187 @@
+/**
+ * C++ program to sort a file data using external sorting
+ * External sorting is required when the data being sorted do not fit into the
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:3]: please follow the documentation guidelines 
read the contribution.md file to see the proper header comment block each file must be accompanied with
```cpp
/**
 * @file
 * @brief < a brief description of the algorithm with a link to one source of the algorithm>
 * @details <a more detailed description of the algorithm
 *
 * @author [yourname](link to your github)
 */
 ````

File: sorting/external_sort.cpp [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+ * memory (RAM)
+ */
+
+#include <algorithm>
+#include <cassert>
+#include <fstream>
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:9]:
```cpp
#include <someheader.h> /// <why some header was included>
````
please follow this pattern for documenting the includes in your algorithm

File: sorting/external_sort.cpp [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+#include <iostream>
+#include <random>
+#include <string>
+
+using namespace std;
+
+// buff data which fit in memory to sort
+constexpr auto BUFF_SIZE = 100000;
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:17]:
```suggestion

constexpr auto BUFF_SIZE = 100000; ///< size of  buffer data for sorting 
```

File: sorting/external_sort.cpp [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+struct ChunkFile {
+    ifstream file;
+    int currentVal;
+    string line;
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:22]: please document this structure
in the format
```cpp
int somevar; ///< <why somevar is here>

File: sorting/external_sort.cpp [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+};
+
+/**
+   Prepare 1M rows data (int) data being sorted and write to text file.
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:26]:
```suggestion
  * @brief Prepares 1M rows data (int) data being sorted and write to text file.
```

File: sorting/external_sort.cpp [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+    for (int i = 0; i < 1000000; i++) {
+        // do random number
+        int randomNumber = distribution(generator);
+        data_file << randomNumber << "\n";
+    }
+    data_file.close();
+
+    cout << "End preparing data" << endl;
+}
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:46]:
```suggestion
void mockData(char* path = "data.txt") {
    ofstream data_file(path , ofstream::trunc);  // overwite

    // use to get a random number
    mt19937 generator(time(0));
    uniform_int_distribution<int> distribution(0, 999999);

    for (int i = 0; i < 1000000; i++) {
        int randomNumber = distribution(generator);
        data_file << randomNumber << "\n";
    }
    data_file.close();
}
```
wouldnt it be better to have a path parameter to specify the path of the file

File: sorting/external_sort.cpp [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+            chunk.clear();
+            line_number = 0;
+        }
+    }
+    // sort the last chunk
+    if (line_number != 0) {
+        sort(chunk.begin(), chunk.end());
+
+        // write sort result to temporary file. use index as file name
+        string temp_file_name = to_string(chunk_file_names.size());
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:88]: this seems highly unnecessary

File: sorting/external_sort.cpp [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    cout << "Merge sorted chunks using k-way merging" << endl;
+
+    // merge all sorted chunks to the final result, using K-Way merging
+    vector<ChunkFile*> chunk_files;
+
+    // change to absolutely path if u want
+    ofstream result_file("sort_result.txt", ofstream::trunc);
+
+    // open chunk files and read the first value
+    for (string f_name : chunk_file_names) {
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:107]:
```suggestion
    for (string filename : chunk_file_names) {
```
there isnt a limit to length of variable names please always make sure to use descriptive names

File: sorting/external_sort.cpp [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+        ChunkFile* tmp_chunk_file = new ChunkFile();
+        tmp_chunk_file->file = ifstream(f_name);
+        getline(tmp_chunk_file->file, tmp_chunk_file->line);
+        tmp_chunk_file->currentVal = stoi(tmp_chunk_file->line);
+        chunk_files.push_back(tmp_chunk_file);
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:112]:
```suggestion
        ChunkFile* tmp_chunk_file = new ChunkFile();
        tmp_chunk_file->file = ifstream(filename);
        getline(tmp_chunk_file->file, tmp_chunk_file->line);
        tmp_chunk_file->currentVal = stoi(tmp_chunk_file->line);
        chunk_files.push_back(tmp_chunk_file);
```

File: sorting/external_sort.cpp [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+    cout << "Test result: OK" << endl;
+}
+
+int main() {
+    mockData();
+    externalSort("data.txt");
+    test();
+    return 0;
+}
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:187]:
```suggestion
}

```

File: sorting/external_sort.cpp [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+
+    // do compare
+    assert(mock_data == sorted_data);
+
+    cout << "Test result: OK" << endl;
+}
+
+int main() {
+    mockData();
+    externalSort("data.txt");
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:184]:
```suggestion
```
you could definitely move this to test

File: sorting/external_sort.cpp [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+        remove(temp_file.c_str());
+    }
+}
+
+/**
+ * Test the sort result. This method only works when the mock data fit in
+ * memory. (in most real cases, it doesn't)
+ */
+void test() {
+    cout << "Testing" << endl;
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:156]:
```suggestion
```

File: sorting/external_sort.cpp [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:157]:
```suggestion
    mockData(); // making test data for testing
    externalSort("data.txt"); // sorting that to be used in the algorithm
    
```

File: sorting/external_sort.cpp [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+    string line;
+
+    // read all mock data
+    vector<int> mock_data;
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:161]:
```suggestion
    vector<int> mock_data{};
```
initialize your variables

File: sorting/external_sort.cpp [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+    ifstream mock_file("data.txt");
+    while (getline(mock_file, line)) {
+        mock_data.push_back(stoi(line));
+    }
+    // sort mock data by built-in function
+    sort(mock_data.begin(), mock_data.end());
+
+    // read all sorted data
+    vector<int> sorted_data;
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:170]:
```suggestion
    vector<int> sorted_data{};
```

File: sorting/external_sort.cpp [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+    // do compare
+    assert(mock_data == sorted_data);
+
+    cout << "Test result: OK" << endl;
+}
+
+int main() {
+    mockData();
+    externalSort("data.txt");
+    test();
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:185]:
```suggestion
    test(); // run self-test implementation
```

File: sorting/external_sort.cpp [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+
+    cout << "End preparing data" << endl;
+}
+
+void externalSort(char* data_file) {
+    ifstream input_file(data_file);
+
+    string line;
+    int line_number = 0;
+    vector<int> chunk;
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:53]:
```suggestion
    vector<int> chunk{};
```

File: sorting/external_sort.cpp [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+    vector<string> chunk_file_names;
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:54]:
```suggestion
    vector<string> chunk_file_names{};
```

File: sorting/external_sort.cpp [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+        chunk.push_back(stoi(line));
+        if (line_number == BUFF_SIZE) {
+            // if have enough buffer data then do sort the chunk
+            // note: i am using a built-in <algorithm> sort function,
+            // you can use any other in-memory sort algorithms such as bubble
+            // sort or quick sort
+            sort(chunk.begin(), chunk.end());  // asc
+
+            // write sort result to temporary file. use index as file name
+            string temp_file_name = to_string(chunk_file_names.size());
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:71]: but you havent assigned chunk file names anything yet

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: sorting/external_sort.cpp (hunk 7 above)
Comment (bls-lehoai) [reply sorting/external_sort.cpp:94]: @realstealthninja sorry i don't get it. could u give me more details on what is unnecessary?

File: sorting/external_sort.cpp (hunk 19 above)
Comment (bls-lehoai) [reply sorting/external_sort.cpp:77]: @realstealthninja 
I don't think we need to assign chunk_file_names cuz it takes the default value for the first time. i suppose the first chunk file name is "0".

File: sorting/external_sort.cpp [hunk 20]
```diff
@@ -0,0 +1,194 @@
...
+ */
+
+#include <algorithm>  // for in-memory sorting
+#include <cassert>    // for testing
+#include <fstream>    // for read/write data file
+#include <iostream>   // for I/O operations
+#include <random>     // for generate a random number
+#include <string>     // for string operations (read file line by line)
+
+using namespace std;
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:18]:
```suggestion

```
avoid using namespace std;

File: sorting/external_sort.cpp [hunk 21]
```diff
@@ -0,0 +1,194 @@
...
+    uniform_int_distribution<int> distribution(0, 999999);
+
+    for (int i = 0; i < 1000000; i++) {
+        // do random number
+        int randomNumber = distribution(generator);
+        data_file << randomNumber << "\n";
+    }
+    data_file.close();
+
+    cout << "End preparing data" << endl;
```
Comment (realstealthninja) [inline sorting/external_sort.cpp:51]:
```suggestion
    std::cout << "End preparing data" << std::endl;
```

File: sorting/external_sort.cpp (hunk 7 above)
Comment (realstealthninja) [reply sorting/external_sort.cpp:94]: im sorry. i should have clarified. i dont understand the use of writing result to a temporary file

**Review by realstealthninja — CHANGES_REQUESTED** [review]
thank you so much for your contribution! :heart:

File: sorting/external_sort.cpp (hunk 7 above)
Comment (bls-lehoai) [reply sorting/external_sort.cpp:94]: @realstealthninja 
It's simply a part of the algorithm, to adapt for data is not fit in memory. Writing the sorted chunk results to temporary files is phase 1 of the algorithm, then using K-Way merging to merge all sorted chunks into a final sort result (phase 2).

```string temp_file_name = to_string(chunk_file_names.size());```

the **temp_file_name** is used to store sorted temporary file names (then pushed to a vector), to use for phase 2.

File: sorting/external_sort.cpp (hunk 20 above)
Comment (bls-lehoai) [reply sorting/external_sort.cpp:18]: @realstealthninja 
sorry, why we can't use namespace std?

**Comment by bls-lehoai** [comment]
@realstealthninja there are anything else have to do?

File: sorting/external_sort.cpp (hunk 21 above)
Comment (bls-lehoai) [reply sorting/external_sort.cpp:51]: i don't think there are any problems when using namespace std to reduce source code

**Review by bls-lehoai — COMMENTED** [review]
Is there anything that has to change?

**Comment by lehoai** [comment]
well, at least can u tell me what is wrong with this pull request?

**Pull request closed without merging**